### PR TITLE
[JIT] Allow mutated values as functional graph inputs

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35421 [JIT] Incremental updates to Alias Db in Mutation Remover pass
* **#33297 [JIT] Allow mutated values as functional graph inputs**

Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

Differential Revision: [D20607175](https://our.internmc.facebook.com/intern/diff/D20607175)

### CODE REVIEW

File: torch/csrc/jit/passes/create_functional_graphs.cpp [hunk 1]
```diff
@@ -159,11 +157,7 @@ struct FunctionalGraphSlicer {
     }
 
     // mutated_values_ already populated with inputs to this node
```
Comment (suo) [inline torch/csrc/jit/passes/create_functional_graphs.cpp]: don't need this comment anymore

File: torch/csrc/jit/ir/alias_analysis.h [hunk 2]
```diff
@@ -80,6 +80,9 @@ class AliasDb {
   // Do any nodes write to an alias set inputed/outputed by `n`?
   TORCH_API bool hasWriters(const Node* n) const;
 
+  // Does the node write to any of its inputs
+  TORCH_API bool writeOccursAtNode(Node* n) const;
```
Comment (suo) [inline torch/csrc/jit/ir/alias_analysis.h]: Can we compose this out of existing pieces we have, so we can avoid adding another method? I see a prviate `getWrites`, which is very similar, and from the name `isMutable` does the same job as well.

**Review by suo — APPROVED** [review]
the logic seems reasonable to me, just minor API comments

File: torch/csrc/jit/ir/alias_analysis.h (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/ir/alias_analysis.h]: yea, i think this was accidentally left in the stack. as you pointed it out it reimplements `writesToAlias`

**Pull request closed without merging**